### DISCUSSION TITLE
Osu! Lazer single and multi threaded performance.

### DISCUSSION THREAD

**Discussion by Taitranz** [ppy/osu#18814] [Q&A] [answered]
### Type

Performance

### Bug description

On Osu! Lazer, the performance of multithread and singled thread appears to be the same.
Also changing the frame limiter does not affect the draws. Fps for multithreaded appears to be "capped" at 600 and for singled thread it is roughly 700fps. I do understand getting more than 1000fps shouldn't be the expectation but on the previously builds of Osu! Lazer it was possible. The decrease in FPS makes the latency very noticeable on higher refresh rates. 
Sorry if I did not structure this correctly.

### Screenshots or videos

https://www.youtube.com/watch?v=o0LtWX3i8ho

### Version

Osu! Lazer 2022.621.0

### Logs

[performance.log](https://github.com/ppy/osu/files/8963272/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/8963273/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/8963274/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/8963275/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/8963276/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/8963277/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8963278/updater.log)
[database.log](https://github.com/ppy/osu/files/8963279/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/8963280/legacy-ipc.log)

**Comment by peppy** [comment]
Have you tried running the latency certifier? Does it run at higher fps? What level can you get to?

**Reply by Taitranz** [reply]
I am able to complete level 6 with extreme consistency, and level 7 semi-consistency. The latency is low when the game runs at 600fps, but the issue is when the framerate falls below 600.

**Reply by peppy** [reply]
And the fps while running the latency certifier is high (as in always 1000fps or greater), correct?

**Reply by Taitranz** [reply]
No, the framerate of the latency certifier is 600-800fps

**Reply by peppy** [reply]
That sounds *very* weird. I can only check remotely at the moment but the frame rate for me also looks lower than expected.

@smoogipoo are you able to check whether your overall draw fps has dropped in the most recent release as well? I'm seeing 1400fps on `2022.528.0` (the version this user tested, haven't tried other released in-between)  at the main menu (audio paused, osu logo idle and large), vs 520fps on the latest release.

**Reply by peppy** [reply]
@Taitranz did you test each of the versions after `.528`? Would be useful to know when your fps starts to dip below 1000.

**Reply by Taitranz** [reply]
I will test and specify the version soon, I am having some internet problems at the moment sorry.

**Reply by Taitranz** [reply]
Build 2022.612.0 is when the performance is reduced
Build 2022.605.1 framerate is over 1000+

**Reply by smoogipoo** [reply]
I can't seem to be able to repro. I've tested main menu (as @peppy), song select, and gameplay, in both multi- and single-threaded modes, and have about the same update and draw FPS on both latest master and 2022.528.0.

I'm testing with `host.AllowBenchmarkUnlimitedFrames = true` on master.

**Reply by peppy** [reply]
I'll re-test when I'm local to my PC. Hard to tell if there is actually an issue or not.

**Reply by Flamiii** [reply]
I've been experiencing lag spikes and microstutters in lazer for the past ~2 weeks and I genuinely thought that it was something on my end; however after testing both the latest release of lazer as well as a local release of 2022.0605.1, I can safely say that I get significantly higher draw FPS on 2022.0605.1 when compared to the latest release. 

On 2022.0605.1, while watching a replay with the same map, skin, and settings, I get around ~600 draw FPS (rough estimate, fluctuates a lot).
On the latest release, I only get around ~400 draw FPS.

**Reply by peppy** [reply]
Lag spikes are not going to be related to this thread but thanks for the data point.

I can confirm I can repro this issue and will look further into it from my end.

**Reply by Taitranz** [reply]
Lag spikes is likely external. Thanks for your time

**Comment by peppy** [comment]
Also please run a previous release you believe to have been getting better FPS on and add a video. You can try a previous release by running an old installer from the releases page. Note it will update back to the latest release next time you close the game.

I believe your fps reduction is being caused by external factors (probably something else running on your system) and this would help in confirming that theory.

**Reply by Taitranz** [reply]
Hi, This is a video of the performance of Build 2022.528.0, Thank you for your time
 https://www.youtube.com/watch?v=iAWGBNcLnp0

**Reply by Taitranz** [reply]
I have two friends that use Lazer exclusively and they both experience the same performance decrease. I am not able to provide a video of their game though. I believe our computer system specification is not the limiting factor in the performance decrease

**Reply by peppy** [reply]
I think this is going to be related to the change made to HUD hiding method. Will need to do some testing but that's the only change made around this period which could affect gameplay draw frame rates.

That said, you don't have anything on the HUD hidden so now too sure yet.

Please answer the more important question I asked above https://github.com/ppy/osu/discussions/18814#discussioncomment-3006862 in the mean time. At the frame rates you're showing you should not be feeling any latency or issues, so I'd want to better understand that first.

**Comment by Loreos7** [comment]
If your game can't be run exclusive fullscreen then this could be the thing. I have the same problem but fps are the same (~250fps). I believe this is because of windows10's fullscreen optimization which can't be turned off due to absence of compatibility mode in lazer.

**Reply by Taitranz** [reply]
I believe this is not the same problem, I have tried this suggestion and it results in the same performance. Thank you

**Comment by peppy** [comment, marked as answer]
This will be resolved via ppy/osu-framework#5267. Thanks for all the diagnostics provided in this thread.

**Comment by Flamiii** [comment]
After testing the latest version of lazer (2022.628.0) and a build from an older version of lazer (2022.605.1), I still seem to get ~150-200 lower draw FPS at any given time on the new version of lazer when compared to the older version. With the 2x refresh rate frame limiter (288fps) that I use, this ends up being ~10% additional CPU/GPU usage for me on the latest version of lazer.

**Reply by peppy** [reply]
If you're seeing a different at anything that is not "basically unlimited" it is likely a different issue. If it's only during gameplay and you use a hidden HUD, it's going to be that for sure.

Also, even if frames are lower, the new update will give you lower latency guaranteed. Frame count is not everything.

**Reply by Loreos7** [reply]
My input latency in single threaded is ~1.8-1.9 ms but in multi its ~2.6-2.7 ms (with 4x rate limiter). Is this how it should be?

**Reply by Taitranz** [reply]
The latest build has less frames then 2022.605.1 but it seems to be less latency than 2022.621.0. This is most definitely a improvement.
Thank you
